drivers: clock_control: stm32h7 fix high frequency setting

lochej opened the pull request:
Fixes #27212 by setting the APBx dividers to maximum
prior to configuring the PLL as clock source.

**Fix related updates:**
They are then set to correct values once the PLL
has been configured.

Prevents going over the limits of APBx clocks when
choosing the PLL as system clock source for
high frequencies (close to 480MHz) which crashes
the MCU clock system.

**Reliability and coherence updates:**
Flash latency setting has been reworked to comply
with the adviced method from STM32 reference manuals
and align with common stm32 method.

**UX updates:**
Preprocessor guard system prevents the user from 
choosing invalid clock configurations.

**HSI divider support** is now added for STM32H7 series
and fully integrated in the clock check procedures.
Configurable with:
**CONFIG_CLOCK_STM32_HSI_DIVISOR** set to 1 by default 
to not break existing boards configurations.

Signed-off-by: Jeremy LOCHE <lochejeremy@gmail.com>

lochej committed:
drivers: clock_control: stm32h7 fix high frequency setting

Fixes #27212 by setting the AHB/APBx dividers
prior to configuring the PLL as clock source.

Prevents going over the limits of APBx clocks when
choosing the PLL as system clock source for
high frequencies (close to 480MHz)

Signed-off-by: Jeremy LOCHE <lochejeremy@gmail.com>

ABOSTM reviewed (approved):
LGTM

Nukersson commented:
Do I understand it right: The settings like `D1CPRE` will be changed to some proper value, no matter, what KConfig setting is?

Please do not merge. I would like test this PR today evening.

lochej commented:
> 
> 
> Do I understand it right: The settings like `D1CPRE` will be changed to some proper value, no matter, what KConfig setting is? Or it affects only APB/AHB prescaler?
> 
> Please do not merge. I would like test this PR today evening.

If the user select high clock frequencies, he must select correct D1CPRE, HPRE, D2PPRE1 ... dividers to comply with the maximum allowable frequency of each bus. This PR makes sure that the dividers are set before applying a selecting the sysclk source. It ensures that reset configuration (setting prescalers to 1) doesn't cause too high frequencies on APBx buses which prevents the system from working properly.

lochej committed:
drivers: clock_control: stm32h7 fixed inverted assertion

Fixed wrong usage of assertions.
Assertions should check that the value is in range and
not out of range.

Signed-off-by: Jeremy LOCHE <lochejeremy@gmail.com>

lochej committed:
drivers: clock_control: stm32h7 add preprocessor clock check

Adds preprocessor clock feasibility check to avoid setting
too high clocks to SYSCLK,AHB,APBx buses.

Also checks if CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
matches with the desired clock configuration by
the M7 core.

Signed-off-by: Jeremy LOCHE <lochejeremy@gmail.com>

lochej commented:
@Nukersson and @ABOSTM can one of you try this PR on his board ? 

I have managed to fix the clock config and produce up to 480Mhz for the system. This is with HSE/HSI/CSI as PLL source clock so this PR should fix all of the clock issues that were introduced by #26980 .

I have added some preprocessor checks to validate the clock setting at compile time, I would love some review from you on this 😄

Nukersson commented:
@lochej Nice work with preprocessor. It is known, that `((uint32_t)value)` may break preprocessor's `>`/`<` comparison or other way around. I had same problem  in one project.

To your prepocessor test:
1. I would add lower frequency range tests.
2. Are the clock calculations (like main CLK, APB/AHB CLK) same for all Fx, Lx, Hx, WB series? If not, they need to hold only for tested (H7 in our case) series e.g. be placed in corresponding defines. But if this driver is only for H7 series, then it is okay. ( Can't check it currently from my smartphone ).

ABOSTM commented:
@lochej thanks fro this update,
I tested on my nucleo_h743zi2, I but face some strange results:
With PLL 480 (HSE) hello_world sample works fine.
With PLL 480 (HSI) hello_world sample doesn't work (software is running but nothing display on console; usart broken ??)
but it works with PLL 480 (HSI) with blinky sample (led is blinking at the expected frequency)
Did you test hello_world sample ? is it working on your side ?

lochej commented:
> 
> 
> @lochej Nice work with preprocessor. It is known, that `((uint32_t)value)` may break preprocessor's `>`/`<` comparison or other way around. I had same problem in one project.
> 
> To your prepocessor test:
> 
>     1. I would add lower frequency range tests.
> 
>     2. Are the clock calculations (like main CLK, APB/AHB CLK) same for all Fx, Lx, Hx, WB series? If not, they need to hold only for tested (H7 in our case) series e.g. be placed in corresponding defines. But if this driver is only for H7 series, then it is okay. ( Can't check it currently from my smartphone ).

The defines are only used in the clock_stm32_ll_h7.c file 😄

lochej commented:
> 
> 
> @lochej thanks fro this update,
> I tested on my nucleo_h743zi2, I but face some strange results:
> With PLL 480 (HSE) hello_world sample works fine.
> With PLL 480 (HSI) hello_world sample doesn't work (software is running but nothing display on console; usart broken ??)
> but it works with PLL 480 (HSI) with blinky sample (led is blinking at the expected frequency)
> Did you test hello_world sample ? is it working on your side ?

@ABOSTM  I have tested using the ```hello_world``` sample but with one slight modification to print the system clock:
```c
#include <zephyr.h>
#include <sys/printk.h>

void main(void)
{
	printk("Hello World! %s\tclk:%d\n", CONFIG_BOARD,CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC);
}
```
Here are the clock configurations I tested working on my Nucleo_H743ZI:
```
# CSI PLL sysclk 480MHz
CONFIG_CLOCK_STM32_PLL_SRC_CSI=y
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=480000000
CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y
CONFIG_CLOCK_STM32_PLL_M_DIVISOR=1
CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER=240
CONFIG_CLOCK_STM32_PLL_P_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_Q_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_R_DIVISOR=2
CONFIG_CLOCK_STM32_D1CPRE=1
CONFIG_CLOCK_STM32_HPRE=2
CONFIG_CLOCK_STM32_D2PPRE1=2
CONFIG_CLOCK_STM32_D2PPRE2=2
CONFIG_CLOCK_STM32_D1PPRE=2
CONFIG_CLOCK_STM32_D3PPRE=2
```
```
# HSI PLL sysclk 480MHz
CONFIG_CLOCK_STM32_PLL_SRC_HSI=y
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=480000000
CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y
CONFIG_CLOCK_STM32_PLL_M_DIVISOR=4
CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER=60
CONFIG_CLOCK_STM32_PLL_P_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_Q_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_R_DIVISOR=2
CONFIG_CLOCK_STM32_D1CPRE=1
CONFIG_CLOCK_STM32_HPRE=2
CONFIG_CLOCK_STM32_D2PPRE1=2
CONFIG_CLOCK_STM32_D2PPRE2=2
CONFIG_CLOCK_STM32_D1PPRE=2
CONFIG_CLOCK_STM32_D3PPRE=2
```

Please note the HSE_BYPASS set to **n** because my board is split from its ST-LINKv2-1 so HSE is provided by an external 8MHz crystal that I soldered on the board instead.
```
# HSE PLL sysclk 480MHz
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=480000000
CONFIG_CLOCK_STM32_PLL_SRC_HSE=y
CONFIG_CLOCK_STM32_HSE_CLOCK=8000000
CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y
CONFIG_CLOCK_STM32_HSE_BYPASS=n
CONFIG_CLOCK_STM32_PLL_M_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER=240
CONFIG_CLOCK_STM32_PLL_P_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_Q_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_R_DIVISOR=2
CONFIG_CLOCK_STM32_HPRE=2
CONFIG_CLOCK_STM32_D1CPRE=1
CONFIG_CLOCK_STM32_D2PPRE1=2
CONFIG_CLOCK_STM32_D2PPRE2=2
CONFIG_CLOCK_STM32_D1PPRE=2
CONFIG_CLOCK_STM32_D3PPRE=2
```

In every case, the USART show the following message:
(This is a custom board definition to change the clock configs in the *defconfig file)
```
*** Booting Zephyr OS build zephyr-v2.3.0-1376-g0562f14bc6ef  ***
Hello World! nucleo_h743zi_custom       clk:480000000
```

ABOSTM commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -449,14 +547,6 @@ static int stm32_clock_control_init(struct device *dev)
-
 	/* Set FLASH latency */
 	/* AXI clock is SYSCLK / HPRE */
 	LL_SetFlashLatency(get_bus_clock(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC,
```
There is a potential issue with Flash latency: 
When clock configuration increases Flash frequency (=AXI clock = HCLK), latency should be set first.
When frequency is decreased, latency should be set after clock configuration.
Issue is probably not visible because after reset default latency is set the max value.
You can check the way it is done for other series in 'clock_stm32_ll_common.c'

lochej replied:
Exactly, in this case we are only tweaking down the flash latency as reset configuration sets it to the maximum value for H7 series (7 wait states).

So it cannot be an issue, can it ?

ABOSTM replied:
It is not an issue today, ... but in the future ...
Nevertheless, I would prefer to keep coherence with other series which are tackling with this.

lochej replied:
Alright then 👍 , seeing the common stm32 implementation it should be implementable in this PR.
I'll do my best to implement the method correctly (but I bet it won't change code execution in the end).

Nukersson replied:
> (but I bet it won't change code execution in the end)
You're doing it for generations after you LOL

lochej replied:
> > (but I bet it won't change code execution in the end)
> 
> You're doing it for generations after you LOL

Yes that's why you changed my mind, I'm preparing something clean for the future generation 👍

lochej replied:
@Nukersson @ABOSTM check my latest commit to see if this new flash latency setting implementation suits you.

Nukersson replied:
> @Nukersson @ABOSTM check my latest commit to see if this new flash latency setting implementation suits you.

Checked it currently out. Still works fine for me with `help_world` and M7 core on my nucleo-h745zi-q board. I've not tested the flash latency properly. I will just hope, it works fine.

lochej commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -31,6 +31,88 @@
+/* prevents from compare with #if > operator*/
+/* original defines in stm32h7xx_hal_conf.h*/
+#define HSI_FREQ 64000000UL	/* HSI_VALUE ((uint32_t)64000000) */
+#define HSE_FREQ HSE_VALUE	/* HSE_VALUE overridden by the build system without cast to uint*/
```
Using the defines from the HAL is not possible as it prevents the preprocessor to do computation on the symbols.
This is explained in the comment above the HSI_FREQ define.

I have used HSI_FREQ, HSE_FREQ,CSI_FREQ as wrapper to redefine the symbols to they are preprocessor ready.
Else we would have to redefine the symbols in the STM32Cube package to remove the explicit cast.

ABOSTM replied:
ok I got it.

lochej commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -31,6 +31,88 @@
+/* Suppress the cast to uint32_t which */
+/* prevents from compare with #if > operator*/
+/* original defines in stm32h7xx_hal_conf.h*/
+#define HSI_FREQ 64000000UL	/* HSI_VALUE ((uint32_t)64000000) */
```
The explicit cast in this define prevents the preprocessor to do computation with this symbol.

lochej commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -31,6 +31,88 @@
+#define APB1_FREQ_MAX	120000000UL
+#define APB2_FREQ_MAX	120000000UL
+#define APB3_FREQ_MAX	120000000UL
+#define APB4_FREQ_MAX	120000000UL
```
Agree, change done ! 👍

Nukersson replied:
> May be simplified to a unique define like
> #define APB_FREQ_MAX	120000000UL

I would suggest `APBx_FREQ_MAX`

lochej replied:
I also agree with APBx which clearly states that is it for all APB buses, this also inlines with the commit message.

Change done !

Nukersson commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -31,6 +31,90 @@
+/* prevents from compare with #if > operator*/
+/* original defines in stm32h7xx_hal_conf.h*/
+#define HSI_FREQ 64000000UL	/* HSI_VALUE ((uint32_t)64000000) */
+#define HSE_FREQ HSE_VALUE	/* HSE_VALUE overridden by the build system without cast to uint*/
```
The line is over 80 characters. Please test like this (run the script from zephyr folder):
> ./scripts/checkpatch.pl -g HEAD-7

lochej replied:
Done, moved the comments to the previous line.

Nukersson commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -31,6 +31,90 @@
+#endif
+
+/* Given source clock and dividers, computed the output frequency of PLLP */
+#define PLLP_FREQ(pllsrc_freq, divm, divn, divp)	(((pllsrc_freq)*(divn))/((divm)*(divp)))
```
The line is over 80 characters. Please test like this (run the script from zephyr folder):
> ./scripts/checkpatch.pl -g HEAD-7

Following should work:
```
#define PLLP_FREQ(pllsrc_freq, divm, divn, divp)	(((pllsrc_freq)*\
							(divn))/((divm)*(divp)))
```

lochej replied:
For some reason, this still gives the 80 characters warning. Because of TAB I guess.

lochej replied:
I have done this change, lines are now less than 80 characters but checkpatch fails because tab space is set to 8 spaces instead of 4 tab spaces.

Nukersson replied:
> I have done this change, lines are now less than 80 characters but checkpatch fails because tab space is set to 8 spaces instead of 4 tab spaces.

I know you're working with VS Code. Please place following code in `Zephyr.code-workspace` file:
```
{
        "folders": [
        {
            "path": "."
        }
    ],
	"settings": {
		"files.associations": {
			"*.c": "c"
        },
        // The number of spaces a tab is equal to. This setting is overridden
        // based on the file contents when `editor.detectIndentation` is true.
        "editor.tabSize": 8,

        // Insert spaces when pressing Tab. This setting is overriden
        // based on the file contents when `editor.detectIndentation` is true.
        "editor.insertSpaces": false,

        // When opening a file, `editor.tabSize` and `editor.insertSpaces`
        // will be detected based on the file contents. Set to false to keep
        // the values you've explicitly set, above.
		"editor.detectIndentation": false,
		"editor.rulers": [80]
    }
}
```

Then place this file in `./zephyrproject` folder. It should give you proper settings. This file works fine for me and check patch runs properly.

lochej replied:
Thanks for the tip, so zephyr has always been working with 8 tab spaces ?

xhpohanka replied:
AFAIK it is same as in linux kernel. Tab = 8 spaces. Check Linux kernel coding style, it is quite funny reading :)

lochej replied:
The indentation is now fixed 😄

Nukersson replied:
> Thanks for the tip, so zephyr has always been working with 8 tab spaces ?

Yes. It is.

Nukersson commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -31,6 +31,90 @@
+#endif
+
+#if SYSCLK_FREQ != CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
+#error "Calculated CPU clock frequency (SYS clock) for M7 core doesn't match CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC"
```
The line is over 80 characters. Please test like this (run the script from zephyr folder):

> ./scripts/checkpatch.pl -g HEAD-7

Following should work:
```
#error "Calculated CPU clock frequency (SYS clock) for M7 core doesn't match"\
	"CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC"
```

lochej replied:
This doesn't show the second line in the preprocessor error, I have to stick to the full line to get the entire message.

Nukersson replied:
> This doesn't show the second line in the preprocessor error, I have to stick to the full line to get the entire message unless we find a way to split the line 🙃

Use this one then:
```
#error "Calculated CPU clock frequency (SYS clock) for M7 core doesn't match \
	CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC"
```

lochej replied:
Looks like this is the way to go:
```
#error "Calculated CPU clock frequency (SYS clock) for M7 core doesn't match \
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC"
```

Gives: 
```
/home/dev/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:120:2: error: #error "Calculated CPU clock frequency (SYS clock) for M7 core doesn't match CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC"
  120 | #error "Calculated CPU clock frequency (SYS clock) for M7 core doesn't match \
      |  ^~~~~
```

Having the tab indent gives:
```
#error "Calculated CPU clock frequency (SYS clock) for M7 core doesn't match \
	CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC"
```
```
/home/dev/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:120:2: error: #error "Calculated CPU clock frequency (SYS clock) for M7 core doesn't match     CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC"
  120 | #error "Calculated CPU clock frequency (SYS clock) for M7 core doesn't match \
      |  ^~~~~
```

KozhinovAlexander reviewed (changes requested):
@lochej @ABOSTM Proposed fix works with my **nucleo_h745zi_q** board with default settings placed in current master branch of Zephyr.

It also works with following settings from above:

1. **CSI PLL sysclk 480MHz**
```
# CSI PLL sysclk 480MHz
CONFIG_CLOCK_STM32_PLL_SRC_CSI=y
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=480000000
CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y
CONFIG_CLOCK_STM32_PLL_M_DIVISOR=1
CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER=240
CONFIG_CLOCK_STM32_PLL_P_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_Q_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_R_DIVISOR=2
CONFIG_CLOCK_STM32_D1CPRE=1
CONFIG_CLOCK_STM32_HPRE=2
CONFIG_CLOCK_STM32_D2PPRE1=2
CONFIG_CLOCK_STM32_D2PPRE2=2
CONFIG_CLOCK_STM32_D1PPRE=2
CONFIG_CLOCK_STM32_D3PPRE=2
```

2. **HSI PLL sysclk 480MHz**
```
# HSI PLL sysclk 480MHz
CONFIG_CLOCK_STM32_PLL_SRC_HSI=y
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=480000000
CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y
CONFIG_CLOCK_STM32_PLL_M_DIVISOR=4
CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER=60
CONFIG_CLOCK_STM32_PLL_P_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_Q_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_R_DIVISOR=2
CONFIG_CLOCK_STM32_D1CPRE=1
CONFIG_CLOCK_STM32_HPRE=2
CONFIG_CLOCK_STM32_D2PPRE1=2
CONFIG_CLOCK_STM32_D2PPRE2=2
CONFIG_CLOCK_STM32_D1PPRE=2
CONFIG_CLOCK_STM32_D3PPRE=2
```

3. **HSE PLL sysclk 480MHz**
```
# HSE PLL sysclk 480MHz
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=480000000
CONFIG_CLOCK_STM32_PLL_SRC_HSE=y
CONFIG_CLOCK_STM32_HSE_CLOCK=8000000
CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y
CONFIG_CLOCK_STM32_HSE_BYPASS=y
CONFIG_CLOCK_STM32_PLL_M_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER=240
CONFIG_CLOCK_STM32_PLL_P_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_Q_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_R_DIVISOR=2
CONFIG_CLOCK_STM32_HPRE=2
CONFIG_CLOCK_STM32_D1CPRE=1
CONFIG_CLOCK_STM32_D2PPRE1=2
CONFIG_CLOCK_STM32_D2PPRE2=2
CONFIG_CLOCK_STM32_D1PPRE=2
CONFIG_CLOCK_STM32_D3PPRE=2
```

_Note:_ `CONFIG_CLOCK_STM32_HSE_BYPASS=y` - were set in the last setting! Since my board uses 8MHz from ST-LINK onboard.

The `hello_world` sample project wer compiled for this test. The minimum console shows expected message.

@lochej Thank you for hot fix!

Below you can see some small changes, I would like you to add.

lochej committed:
drivers: clock_control: stm32h7 change set flash latency

Updated flash latency setting to comply with
common stm32 implementation and datasheet.

Signed-off-by: Jeremy LOCHE <lochejeremy@gmail.com>

xhpohanka commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -31,6 +31,96 @@
+
+/* Choose PLL SRC */
+#ifdef CONFIG_CLOCK_STM32_PLL_SRC_HSI
+#define PLLSRC_FREQ  HSI_FREQ
```
`HSI_FREQ` can be divided by `HSIDIV`. I have it such in one of my projects. In such case the checks for `SYSCLK_FREQ > SYSCLK_FREQ_MAX` fails by mistake.

xhpohanka replied:
Now I see you probably do not have there support for `HSIDIV != 1` at all. I found that also have it hardcoded in my project...

lochej replied:
Yes you are right, HSI can be divided.

What symbol to you use to provide the HSIDIV value ? So I can add it to the driver.

xhpohanka replied:
We probably need to include a new value to Kconfig for HSIDIV. I had there just hardcoded line setting that divider to 8, so for me the HSI freq was always 8MHz. When I was trying your PR I just hit your guards, that are right for `HSIDIV == 1`.

lochej replied:
To add the HSI divisor configuration I suggest: 
Name:
1. CLOCK_STM32_HSI_DIVISOR : 
2. Parameters: default to 1, range: 1 8, allowed values: 1 2 4 8

xhpohanka replied:
No problem for me. Please note that I'm not a Zephyr developer/maintainer. I just use it on STM32H7 platform so your patches are handy for me - thank you. I had also my changes on similar topic, but never had time to polish them and make public.

lochej replied:
Check out the latest commit, I have tested this implementation with:
```
# HSI divided to 8MHz test
CONFIG_CLOCK_STM32_SYSCLK_SRC_HSI=y
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=8000000
CONFIG_CLOCK_STM32_HSI_DIVISOR=8
```
and also from PLL at 480MHz:
```
# HSI divided to 8MHz + PLL 480MHz test
CONFIG_CLOCK_STM32_PLL_SRC_HSI=y
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=480000000
CONFIG_CLOCK_STM32_HSI_DIVISOR=8
CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y
CONFIG_CLOCK_STM32_PLL_M_DIVISOR=1
CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER=120
CONFIG_CLOCK_STM32_PLL_P_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_Q_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_R_DIVISOR=2
CONFIG_CLOCK_STM32_D1CPRE=1
CONFIG_CLOCK_STM32_HPRE=2
CONFIG_CLOCK_STM32_D2PPRE1=2
CONFIG_CLOCK_STM32_D2PPRE2=2
CONFIG_CLOCK_STM32_D1PPRE=2
CONFIG_CLOCK_STM32_D3PPRE=2
```

xhpohanka replied:
I have tested it on my board. Works fine.

lochej replied:
This addition will be added to PR description.

The code will have to be reviewed by approved reviewers to see if it can be embedded in this PR.
@ABOSTM I would be glad to here what you think 👍

Nukersson replied:
> I have tested it on my board. Works fine.

Please name the board and/or reference to it if it is already available in Zephyr's main repo.

Nukersson replied:
> To add the HSI divisor configuration I suggest: 
> Name:
> 1. CLOCK_STM32_HSI_DIVISOR : 
> 2. Parameters: default to 1, range: 1 8, allowed values: 1 2 4 8

I think DIV instead of DIVISOR will be better, cause shorter.

xhpohanka replied:
> > I have tested it on my board. Works fine.
> 
> Please name the board and/or reference to it if it is already available in Zephyr's main repo.

It's a custom board, not suitable for main repo unfortunately.

Nukersson replied:
> No problem for me. Please note that I'm not a Zephyr developer/maintainer. I just use it on STM32H7 platform so your patches are handy for me - thank you. I had also my changes on similar topic, but never had time to polish them and make public.
> BTW so far all your patches (including ETH) work on my custom board with STM32H750VBT6.

Nice to hear, you can use it in your project. I hope, you can profit from the project more and more with it's growth.

xhpohanka replied:
`CLOCK_STM32_HSI_DIVISOR` I maybe left the DIVISOR suffix here, to be in same manner as other existing values like `CLOCK_STM32_PLL_M_DIVISOR`.

lochej replied:
> I think DIV instead of DIVISOR will be better, cause shorter.
I agree.
But I chose DIVISOR to align with the PLL symbols:
```
CLOCK_STM32_PLL_M_DIVISOR
CLOCK_STM32_PLL_R_DIVISOR
CLOCK_STM32_PLL_Q_DIVISOR
```

lochej committed:
drivers: clock_control: stm32h7 add HSI divisor support

Add HSI divisor support for clock tree configuration.
Removed HSI calibration trimming to comply with
common STM32 implementation and use reset default
configuration instead.

Signed-off-by: Jeremy LOCHE <lochejeremy@gmail.com>

ABOSTM commented on drivers/clock_control/Kconfig.stm32h7:
```diff
@@ -5,6 +5,16 @@
+	default 1
+	range 1 8
+	help
+	  HSI Divisor to divide HSI base frequency value
```
Could add a dependence on (CONFIG_CLOCK_STM32_PLL_SRC_HSI || CONFIG_CLOCK_STM32_SYSCLK_SRC_HSI)

lochej replied:
Dependance added 😄

ABOSTM commented on drivers/clock_control/Kconfig.stm32h7:
```diff
@@ -5,6 +5,16 @@
+config CLOCK_STM32_HSI_DIVISOR
+	int "HSI Divisor"
+	default 1
+	range 1 8
```
I am not expert in Kconfig, but I wonder whether there is a way list allowed value (and avoid user to set 3 for example)

lochej replied:
> 
> 
> I am not expert in Kconfig, but I wonder whether there is a way list allowed value (and avoid user to set 3 for example)

Me neither 🙃 , I'll investigate on this and make the change if this is an available option.

Nukersson replied:
@lochej  You can at least add following help on the new line right after `range` keyword:
```
help Accepted values are: 1, 2, 4, 8
```

ABOSTM reviewed (changes requested):
Difficult to make test: each time I did it, a new commit comes :-)
I'm joking but really appreciate your great job. Thanks a lot.
I tested with my nucleo_h743zi2 and it works fine with PLL (HSE) 480MHz
It also worked with PLL (HSI, HSI_DIV8) 480MHz (but I needed to workaround uart issue I face yesterday when using HSI: for now I removed HSI calibration setting but I will investigate further as it is a separate issue).
I have a remaining minor remark.

lochej commented:
> 
> 
> Difficult to make test: each time I did it, a new commit comes :-)
> I'm joking but really appreciate your great job. Thanks a lot.
> I tested with my nucleo_h743zi2 and it works fine with PLL (HSE) 480MHz
> It also worked with PLL (HSI, HSI_DIV8) 480MHz (but I needed to workaround uart issue I face yesterday when using HSI: for now I removed HSI calibration setting but I will investigate further as it is a separate issue).
> I have a remaining minor remark.

I'm actively working on the subject to correct the reviews that's why so many changes 😄.
About the calibration trimming, now that I think about it, it might not be an a good idea to add it the driver if that can cause issues to users. I probably had better removing this configuration if this is the root of your problem.

We'll need help test from your side to see if that's the actual issue.

ABOSTM commented:
@lochej ,
I finally clarify the HSI calibration issue:
As per Reference Manual (RM0433) the HSI Clock trimming depends on chip revision:
* Rev Y : HSITRIM bitfield is in RCC_ICSCR and default value of HSITRIM is 0x20=32
* Rev V : HSITRIM bitfield is in RCC_HSICFGR and default value of HSITRIM is 0x40=64
(see description of RCC_ICSCR and RCC_HSICFGR)

So when we call `LL_RCC_HSI_SetCalibTrimming(32);`   to set default value,
this is not always valid.

But there is no need to implement a complex function:
Knowing that we only wants to set default value which is already the value after reset, this configuration is useless, 
Knowing that zephyr implementation for other STM32 series doesn't perform HSITRIM setting to default value..
So yes I agree to remove both occurrences of this function call LL_RCC_HSI_SetCalibTrimming(32).

lochej commented:
> 
> 
> @lochej ,
> I finally clarify the HSI calibration issue:
> As per Reference Manual (RM0433) the HSI Clock trimming depends on chip revision:
> 
>     * Rev Y : HSITRIM bitfield is in RCC_ICSCR and default value of HSITRIM is 0x20=32
> 
>     * Rev V : HSITRIM bitfield is in RCC_HSICFGR and default value of HSITRIM is 0x40=64
>       (see description of RCC_ICSCR and RCC_HSICFGR)
> 
> 
> So when we call `LL_RCC_HSI_SetCalibTrimming(32);` to set default value,
> this is not always valid.
> 
> But there is no need to implement a complex function:
> Knowing that we only wants to set default value which is already the value after reset, this configuration is useless,
> Knowing that zephyr implementation for other STM32 series doesn't perform HSITRIM setting to default value..
> So yes I agree to remove both occurrences of this function call LL_RCC_HSI_SetCalibTrimming(32).

@ABOSTM Thank you for your research that's very interesting, that can very well explain why you faced problems. I have removed all occurances of LL_RCC_HSI_SetCalibTrimming in the latest commit. 
I hope this solves your issues. Only the tests will tell ! 😄

ABOSTM reviewed (approved):
hello_world passed on nucleo_h743zi2
* PLL(HSE) 96MHz (default)
* PLL(HSE) 480MHz
* PLL(HSI/8) 480MHz

LGTM

Nukersson commented:
> > Difficult to make test: each time I did it, a new commit comes :-)
> > I'm joking but really appreciate your great job. Thanks a lot.
> > I tested with my nucleo_h743zi2 and it works fine with PLL (HSE) 480MHz
> > It also worked with PLL (HSI, HSI_DIV8) 480MHz (but I needed to workaround uart issue I face yesterday when using HSI: for now I removed HSI calibration setting but I will investigate further as it is a separate issue).
> > I have a remaining minor remark.
> 
> I'm actively working on the subject to correct the reviews that's why so many changes 😄.
> About the calibration trimming, now that I think about it, it might not be an a good idea to add it the driver if that can cause issues to users. I probably had better removing this configuration if this is the root of your problem.
> 
> We'll need help test from your side to see if that's the actual issue.

If it is not game stopper here - I would not add this feature in this PR. Instead open another one PR. Too much changes leads to too much possible errors.

Nukersson commented:
> @lochej ,
> I finally clarify the HSI calibration issue:
> As per Reference Manual (RM0433) the HSI Clock trimming depends on chip revision:
> 
> * Rev Y : HSITRIM bitfield is in RCC_ICSCR and default value of HSITRIM is 0x20=32
> * Rev V : HSITRIM bitfield is in RCC_HSICFGR and default value of HSITRIM is 0x40=64
>   (see description of RCC_ICSCR and RCC_HSICFGR)
> 
> So when we call `LL_RCC_HSI_SetCalibTrimming(32);` to set default value,
> this is not always valid.
> 
> But there is no need to implement a complex function:
> Knowing that we only wants to set default value which is already the value after reset, this configuration is useless,
> Knowing that zephyr implementation for other STM32 series doesn't perform HSITRIM setting to default value..
> So yes I agree to remove both occurrences of this function call LL_RCC_HSI_SetCalibTrimming(32).

We can determine H7 SoC revision for it, reading the appropriate value from `DBGMCU` (see RM0399 Rev 3 3349/3528).
REV_ID has only four values [here](https://www.st.com/resource/en/errata_sheet/dm00368411-stm32h742xig-and-stm32h743xig-device-limitations-stmicroelectronics.pdf) (see Table 2. Device variants). Same for other SoC variants like [stm32h745/7](https://www.st.com/resource/en/errata_sheet/dm00530531-stm32h745747xig-and-stm32h755757xi-device-limitations-stmicroelectronics.pdf)

Nukersson commented:
> > @lochej ,
> > I finally clarify the HSI calibration issue:
> > As per Reference Manual (RM0433) the HSI Clock trimming depends on chip revision:
> > ```
> > * Rev Y : HSITRIM bitfield is in RCC_ICSCR and default value of HSITRIM is 0x20=32
> > 
> > * Rev V : HSITRIM bitfield is in RCC_HSICFGR and default value of HSITRIM is 0x40=64
> >   (see description of RCC_ICSCR and RCC_HSICFGR)
> > ```
> > 
> > 
> > So when we call `LL_RCC_HSI_SetCalibTrimming(32);` to set default value,
> > this is not always valid.
> > But there is no need to implement a complex function:
> > Knowing that we only wants to set default value which is already the value after reset, this configuration is useless,
> > Knowing that zephyr implementation for other STM32 series doesn't perform HSITRIM setting to default value..
> > So yes I agree to remove both occurrences of this function call LL_RCC_HSI_SetCalibTrimming(32).
> 
> @ABOSTM Thank you for your research that's very interesting, that can very well explain why you faced problems. I have removed all occurances of LL_RCC_HSI_SetCalibTrimming in the latest commit.
> I hope this solves your issues. Only the tests will tell ! 😄

I think new PR can be proposed for HSI calibration functionality later.

Nukersson commented:
@lochej Can you please put your branch on top of current master and push it? (actually would better to do it every time you push).

You can do it like this:

1. `git checkout master` (or `git checkout origin/master`) - depends on that which works for you.
2. `git pull upstream/master`
3. `git push origin master --force`
4. `git checkout <this branch name>`
5. `git pull origin/master`
6. **agree with merge***
7. `git rebase origin/master`
8. `git push origin <this branch name> --force`

lochej commented:
> 
> If it is not game stopper here - I would not add this feature in this PR. Instead open another one PR. Too much changes leads to too much possible errors.

Removing calibration timing with this PR solves all of the clock issues introduced by #26980 which was the main point for the H7 series.

Calibration trimming is already done properly at reset so we should be fine until a new PR bringing HSI calibration features comes up.

ABOSTM commented:
> @lochej @ABOSTM
> We can determine H7 SoC revision for it, reading the appropriate value from `DBGMCU` (see RM0399 Rev 3 3349/3528).
> REV_ID has only four values [here](https://www.st.com/resource/en/errata_sheet/dm00368411-stm32h742xig-and-stm32h743xig-device-limitations-stmicroelectronics.pdf) (see Table 2. Device variants). Same for other SoC variants like [stm32h745/7](https://www.st.com/resource/en/errata_sheet/dm00530531-stm32h745747xig-and-stm32h755757xi-device-limitations-stmicroelectronics.pdf)

I know, but it add complexity for nothing (setting value which is already the default one). And it is not coherent with other STM32 families.
So I would prefer to remove.

lochej commented:
@ABOSTM @erwango is in vacation until the 10th of august, maybe should we ask for another certified reviewer to have a look at this PR to speed up its integration ? else if the issues brought by first buggy PR is not a showstopper, we can wait.

ABOSTM commented:
> @ABOSTM @erwango is in vacation until the 10th of august, maybe should we ask for another certified reviewer to have a look at this PR to speed up its integration ? else if the issues brought by first buggy PR is not a showstopper, we can wait.

I have no name to propose, so I we need to wait for erwango.

Nukersson commented:
> > @ABOSTM @erwango is in vacation until the 10th of august, maybe should we ask for another certified reviewer to have a look at this PR to speed up its integration ? else if the issues brought by first buggy PR is not a showstopper, we can wait.
> 
> I have no name to propose, so I we need to wait for erwango.

What about @carlescufi @gmarull ?

KozhinovAlexander reviewed (commented):
Please add more information on preprocessor error message. From my side it should be last step (I hope :) ). And great work Jeremy!

Nukersson commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -31,6 +31,97 @@
+#endif
+
+#if SYSCLK_FREQ != CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
+#error "Calculated CPU clock frequency (SYS clock) for M7 core doesn't match \
```
Another staff:

Putting clock settings from below into M4 defconfig does not trigger this error, if `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` will be changed to any arbitrary value.

the settings:
```
# CSI PLL sysclk 240MHz
# CONFIG_CLOCK_STM32_PLL_SRC_HSE=y
CONFIG_CLOCK_STM32_PLL_SRC_HSI=y
# CONFIG_CLOCK_STM32_PLL_SRC_CSI=y

CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y

# CSI n; HSI: n; HSE: y;
# CONFIG_CLOCK_STM32_HSE_BYPASS=n
CONFIG_CLOCK_STM32_HSE_CLOCK=8000000

CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=240000099

# CSI: ?; HSI: 8; HSE: x;
CONFIG_CLOCK_STM32_HSI_DIVISOR=8
# CSI: 4; HSI: 1; HSE 2;
CONFIG_CLOCK_STM32_PLL_M_DIVISOR=1
# CSI: 240; HSI: 60; HSE ;
CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER=60
# CSI: 2; HSI: 1; HSE 2;
CONFIG_CLOCK_STM32_PLL_P_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_Q_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_R_DIVISOR=2

CONFIG_CLOCK_STM32_D1CPRE=1
CONFIG_CLOCK_STM32_HPRE=2
CONFIG_CLOCK_STM32_D2PPRE1=2
CONFIG_CLOCK_STM32_D2PPRE2=2
CONFIG_CLOCK_STM32_D1PPRE=2
CONFIG_CLOCK_STM32_D3PPRE=2
```

ABOSTM replied:
@Nukersson ,
On dual core (H745 , H743) the clock tree is common to both cores. So to avoid conflict, it has been decided on zephyr that only cortex M7 will perform Clock configuration (look at `#if !defined(CONFIG_CPU_CORTEX_M4)` in `stm32_clock_control_init()`).
So M4 will not configure clocks, this is why no check is performed on M4.
But the check will be done by M7, which ensure all frequencies (for both core) are in the proper range.
The point is that cortex M4 zephyr build still need to know the value of CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC .
so there is a constraint: CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC  in M4 defconfig MUST match with M7 HCLK computation (computed from M7 defconfig).
But there is no mean to ensure such coherence between 2 boards within zephyr build system.

Nukersson replied:
Okay. Understand it.

lochej replied:
@ABOSTM is right.
Zephyr doesn't provide this possiblity and I have added a comment before the preprocessor check on why the clock tree check is only performed by the M7 core.

@Nukersson I cannot had the check for clock coherence in M4 core a @ABOSTM said. I tried trust me 😫 
The only thing that could be done is a runtime check but I don't think this would improve the situation much.

At least, the user will be prompt if its M7 clock config are not feasable.

But your idea for Pragma messages is great and could help the user in case of error detection 👍 I'm gonna test that right now 😄

lochej replied:
Gives this kind of outputs:
```
/home/dev/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:124:9: note: #pragma message: SYSCLK_FREQ is: (((((((64000000UL)/(4)))* (120))/((1)*(2))))/(1))
  124 | #pragma message "SYSCLK_FREQ is: " XSTR(SYSCLK_FREQ)
      |         ^~~~~~~
/home/dev/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:125:9: note: #pragma message: CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC is: 480000000
  125 | #pragma message "CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC is: "\
      |         ^~~~~~~
/home/dev/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:127:2: error: #error "Calculated CPU clock frequency (SYS clock) for M7 core doesn't match CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC"
  127 | #error "Calculated CPU clock frequency (SYS clock) for M7 core doesn't match \
      |  ^~~~~
```
Do you still think it can help the user ? Maybe some message indicating to review prescalers/dividers to match CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC whould be better.

Nukersson replied:
> Gives this kind of outputs:
> ```
> /home/dev/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:124:9: note: #pragma message: SYSCLK_FREQ is: (((((((64000000UL)/(4)))* (120))/((1)*(2))))/(1))
>   124 | #pragma message "SYSCLK_FREQ is: " XSTR(SYSCLK_FREQ)
>       |         ^~~~~~~
> /home/dev/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:125:9: note: #pragma message: CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC is: 480000000
>   125 | #pragma message "CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC is: "\
>       |         ^~~~~~~
> /home/dev/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:127:2: error: #error "Calculated CPU clock frequency (SYS clock) for M7 core doesn't match CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC"
>   127 | #error "Calculated CPU clock frequency (SYS clock) for M7 core doesn't match \
>       |  ^~~~~
> ```
> Do you still think it can help the user ? Maybe some message indicating to review prescalers/dividers to match CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC whould be better.

I think `SYSCLK_FREQ` output is good. It is because of fact, that GCC preprocessor stringificiation does not evaluate the value, but just prints the define you proposed. Neverless, the user can see here, that one of the dividers/miltipliers (4, 120, 1, 2, 1) is wrong and how `SYSCLK_FREQ` will be built. This can give the user input, to adapt the dividers/miltipliers better.

lochej replied:
Alright, I'd like to know what others think about this but my suggestion is:

```c
#if SYSCLK_FREQ != CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
#error "Calculated CPU clock frequency (SYSCLK) for M7 core doesn't match \
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC."
/* C preprocessor stringification macros */
#define XSTR(x) STR(x)
#define STR(x) #x
#pragma message "SYSCLK_FREQ with given prescalers is: "\
XSTR(SYSCLK_FREQ)
#pragma message "CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC is: "\
XSTR(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC)
#endif
```

I've moved STR/XSTR macros in the specific error case such that in case of redefinition, the error message is printed before the redefined symbol error (my guess...).

Giving:
```
/home/dev/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:121:2: error: #error "Calculated CPU clock frequency (SYSCLK) for M7 core doesn't match CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC."
  121 | #error "Calculated CPU clock frequency (SYSCLK) for M7 core doesn't match \
      |  ^~~~~
/home/dev/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:126:9: note: #pragma message: SYSCLK_FREQ with given prescalers is: (((((((64000000UL)/(4)))* (120))/((1)*(2))))/(1))
  126 | #pragma message "SYSCLK_FREQ with given prescalers is: "\
      |         ^~~~~~~
/home/dev/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:128:9: note: #pragma message: CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC is: 480000000
  128 | #pragma message "CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC is: "\
```

ABOSTM replied:
My personal though is that the first basic error message was good "Calculated CPU clock frequency (SYSCLK) for M7 core doesn't match CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC."
With this error, user is able to check by himself why `SYSCLK_FREQ != CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC`
and revert computation.
So for me it is a bit too much. But as it is already done (and well done), I won't reject it. 
(By the way it a clever way to pass define value in preprocessor error)

Nukersson commented:
Works fine with following settings on nucleo-h745zi-q board:

1. CSI/HSI/HSE (480MHz) on M7 core with `hello_world` sample project
2. CSI/HSI/HSE (240MHz) on M4 core with `basic/binky` sample project

Please add preprocessor messages (see my last review).

Otherwise LGTM

lochej commented:
> 
> 
> > > @ABOSTM @erwango is in vacation until the 10th of august, maybe should we ask for another certified reviewer to have a look at this PR to speed up its integration ? else if the issues brought by first buggy PR is not a showstopper, we can wait.
> > 
> > 
> > I have no name to propose, so I we need to wait for erwango.
> 
> What about @carlescufi @gmarull ?

@FRASTM could maybe interested

ABOSTM commented:
> @FRASTM could be interested

In vacation too :-(

carlescufi commented:
@Nukersson could you take another look please?
@gmarull do you have the bandwidth to review?

Nukersson commented:
> @Nukersson could you take another look please?
> @gmarull do you have the bandwidth to review?

@carlescufi 
`samples/hello_world` works fine with CSI/HSI/HSE on nucleo_h745zi_q M7 core. I can't test M4 currently - suggest a bug but not from this PR.

The tested clock settings for this PR were:
```
# PLL sysclk 480MHz: (choose HSE, HSI or CSI by uncommenting)
CONFIG_CLOCK_STM32_PLL_SRC_HSE=y
#CONFIG_CLOCK_STM32_PLL_SRC_HSI=y
#CONFIG_CLOCK_STM32_PLL_SRC_CSI=y

# CSI n; HSI: n; HSE: y;
# CONFIG_CLOCK_STM32_HSE_BYPASS=y

CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y
CONFIG_CLOCK_STM32_HSE_CLOCK=8000000
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=480000000

CONFIG_CLOCK_STM32_HSI_DIVISOR=8
CONFIG_CLOCK_STM32_PLL_M_DIVISOR=1
CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER=120
CONFIG_CLOCK_STM32_PLL_P_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_Q_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_R_DIVISOR=2

CONFIG_CLOCK_STM32_D1CPRE=1
CONFIG_CLOCK_STM32_HPRE=2
CONFIG_CLOCK_STM32_D2PPRE1=2
CONFIG_CLOCK_STM32_D2PPRE2=2
CONFIG_CLOCK_STM32_D1PPRE=2
CONFIG_CLOCK_STM32_D3PPRE=2
```

lochej commented:
> > @Nukersson could you take another look please?
> > @gmarull do you have the bandwidth to review?
> 
> @carlescufi
> `samples/hello_world` works fine with CSI/HSI/HSE on nucleo_h745zi_q M7 core. I can't test M4 currently - suggest a bug but not from this PR.
> 
> The tested clock settings for this PR were:
> 
> ```
> # PLL sysclk 480MHz: (choose HSE, HSI or CSI by uncommenting)
> CONFIG_CLOCK_STM32_PLL_SRC_HSE=y
> #CONFIG_CLOCK_STM32_PLL_SRC_HSI=y
> #CONFIG_CLOCK_STM32_PLL_SRC_CSI=y
> 
> # CSI n; HSI: n; HSE: y;
> # CONFIG_CLOCK_STM32_HSE_BYPASS=y
> 
> CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y
> CONFIG_CLOCK_STM32_HSE_CLOCK=8000000
> CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=480000000
> 
> CONFIG_CLOCK_STM32_HSI_DIVISOR=8
> CONFIG_CLOCK_STM32_PLL_M_DIVISOR=1
> CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER=120
> CONFIG_CLOCK_STM32_PLL_P_DIVISOR=2
> CONFIG_CLOCK_STM32_PLL_Q_DIVISOR=2
> CONFIG_CLOCK_STM32_PLL_R_DIVISOR=2
> 
> CONFIG_CLOCK_STM32_D1CPRE=1
> CONFIG_CLOCK_STM32_HPRE=2
> CONFIG_CLOCK_STM32_D2PPRE1=2
> CONFIG_CLOCK_STM32_D2PPRE2=2
> CONFIG_CLOCK_STM32_D1PPRE=2
> CONFIG_CLOCK_STM32_D3PPRE=2
> ```

Could it be due to @ABOSTM PR #26765 that was merged recently ? That suggests that boot for the M4 processor has changed.

Nukersson commented:
> > > @Nukersson could you take another look please?
> > > @gmarull do you have the bandwidth to review?
> > 
> > @carlescufi
> > `samples/hello_world` works fine with CSI/HSI/HSE on nucleo_h745zi_q M7 core. I can't test M4 currently - suggest a bug but not from this PR.
> > 
> > The tested clock settings for this PR were:
> > 
> > ```
> > # PLL sysclk 480MHz: (choose HSE, HSI or CSI by uncommenting)
> > CONFIG_CLOCK_STM32_PLL_SRC_HSE=y
> > #CONFIG_CLOCK_STM32_PLL_SRC_HSI=y
> > #CONFIG_CLOCK_STM32_PLL_SRC_CSI=y
> > 
> > # CSI n; HSI: n; HSE: y;
> > # CONFIG_CLOCK_STM32_HSE_BYPASS=y
> > 
> > CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y
> > CONFIG_CLOCK_STM32_HSE_CLOCK=8000000
> > CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=480000000
> > 
> > CONFIG_CLOCK_STM32_HSI_DIVISOR=8
> > CONFIG_CLOCK_STM32_PLL_M_DIVISOR=1
> > CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER=120
> > CONFIG_CLOCK_STM32_PLL_P_DIVISOR=2
> > CONFIG_CLOCK_STM32_PLL_Q_DIVISOR=2
> > CONFIG_CLOCK_STM32_PLL_R_DIVISOR=2
> > 
> > CONFIG_CLOCK_STM32_D1CPRE=1
> > CONFIG_CLOCK_STM32_HPRE=2
> > CONFIG_CLOCK_STM32_D2PPRE1=2
> > CONFIG_CLOCK_STM32_D2PPRE2=2
> > CONFIG_CLOCK_STM32_D1PPRE=2
> > CONFIG_CLOCK_STM32_D3PPRE=2
> > ```
> 
> Could it be due to @ABOSTM PR #26765 that was merged recently ? That suggests that boot for the M4 processor has changed.

Don't want speculate on it now. But same behavior I can see currently on the master. I will investigate it later. It is much more important now to merge this PR. As a next step I will investigate dual boot support. Otherway we will suck here on creating PR's - bugs - PR's

KozhinovAlexander reviewed (approved):
LGTM
